### ISSUE TITLE
[Feature] automatic trigger for `objectWillChange` publisher

### ISSUE THREAD

**Issue by BrentMifsud** [BrentMifsud/AsyncValue#5]
Remembering to add a `willSet` to all of your async value properties is not all that user friendly. I think we can do better:

## Proposed Solution:

```swift
// this might not be needed with the new swift 5.7 generics. Will investigate. When Xcode 14 is released.
internal protocol AnyAsyncValue {
     var projectedValue: AsyncSequence<Any>
}

internal extension AsyncValue: AnyAsyncValue {
    var projectedValue: AsyncSequence<Any> {
        projectedValue
    }
}

@MainActor open class AsyncObservableObject: ObservableObject {
    private var objectWillChangeTasks: [Task<Void, Never>] = []
    
    init() {
         setupObjectWillChangeTasks()
    }

    deinit {
        for task in objectWillChangeTasks {
            task.cancel()
        }
    }

    private func setupObjectWillChangeTasks() {
        let mirror = Mirror(reflecting: self)

        for child in mirror.children {
              if let asyncValue = child.value as? AnyAsyncValue {
                  let task = Task {
                       for await _ in asyncValue.projectedValue {
                           objectWillChange.send()
                       }
                  }

                  objectWillChangeTasks.append(task)
              }
        } 
    }
}
```

**Comment by jlsiewert** [comment]
Turns out that is actually simple to to once you know how, since that does not seem to be documented anywhere but in the [Swift Evolution Proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0258-property-wrappers.md#referencing-the-enclosing-self-in-a-wrapper-type).

Credit to [SwiftBySundell](https://www.swiftbysundell.com/articles/accessing-a-swift-property-wrappers-enclosing-instance/) for finding and writing about that.

The tldr of the implementation is that the implementation can provide an alternative built with a `static subscript` with a `KeyPath` to the enclosing instance of the wrapped property for class-types. Here, we further restrict that implementation to subclasses of `ObservableObject` that use the (default) `ObservableObjectPublisher`.

I have also changed the associated test case, seems to be working fine.

**Comment by BrentMifsud** [comment]
solved by #6
